### ISSUE TITLE
false positive "redundant section"

### ISSUE THREAD

**Issue by dten** [ndmitchell/hlint#1257]
the following function has the suggestion included however following the hint results in it not compiling

```
weirdEq :: (Int, Int) -> (Int, Int) -> Bool
weirdEq (_, 1) (_, 1) = True
weirdEq a b = ((==) `on`) snd a b
```

```
Suggestion: Redundant section
Found:
  ((==) `on`) snd
Perhaps:
  (==) `on` snd
```

**Comment by ndmitchell** [comment]
I think the suggestion needs brackets to be valid:

```
weirdEq a b = ((==) `on` snd) a b
```

I think the issue here is that HLint thinks you should replace ``((==) `on`)`` with ``(==) `on` snd``, but maintaining the AST. It doesn't realise the thing on the LHS was atomic, the thing on the RHS wasn't, and given the hole it's going in to, the lack of brackets will cause an issue. I think it probably should figure that out and adapt the suggestion to include brackets. What I'm not sure is if it should do so for this particular hint, or more generally in the HLint machinery.
